### ISSUE TITLE
kotlin.UninitializedPropertyAccessException when I use Kotlin with ButterKnife

### ISSUE THREAD

**Issue by sergio11** [JetBrains/kotlin-examples#38]
Hi,

I followed his example and when i run the application I get the following error:

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.example.sergio.todo_list_firebase/com.example.sergio.todo_list_firebase.TodoListActivity}: kotlin.UninitializedPropertyAccessException: lateinit property toolbar has not been initialized
                      at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2325)
                      at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2387)
                      at android.app.ActivityThread.access$800(ActivityThread.java:151)
                      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1303)
                      at android.os.Handler.dispatchMessage(Handler.java:102)
                      at android.os.Looper.loop(Looper.java:135)
```

This is an excerpt of Activity:

```

class TodoListActivity : AppCompatActivity() {

    @BindView(R.id.toolbar)
    lateinit var toolbar: Toolbar
    @BindView(R.id.fab)
    lateinit var fab: FloatingActionButton
    @BindView(R.id.remindersList)
    lateinit var recyclerView: RecyclerView
    private var reminders: MutableList<Reminder> = ArrayList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_todo_list)
        ButterKnife.bind(this)
        setSupportActionBar(toolbar)
        val llm = LinearLayoutManager(this)
        val adapter = RemindersRecyclerViewAdapter(reminders, this)
        recyclerView!!.layoutManager = llm
        recyclerView!!.adapter = adapter

        fab?.setOnClickListener { view -> Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG).setAction("Action", null).show() }
    }
```

I am using:

```
compile 'com.jakewharton:butterknife:8.4.0'
 apt 'com.jakewharton:butterknife-compiler:8.4.0'

kapt {
    generateStubs = true
}

kotlin_version = '1.0.4'
```

Thansk!!

**Comment by developer--** [comment]
+1

**Comment by chpengzh** [comment]
Butterknife is based on apt, but for kotlin, you'd like to use [kapt](https://blog.jetbrains.com/kotlin/2015/05/kapt-annotation-processing-for-kotlin/)

**Comment by chpengzh** [comment]
in kotlin, we'd like to use [delegate property](http://kotlinlang.org/docs/reference/delegated-properties.html) instead of class field

**Comment by chpengzh** [comment]
there's a example for kotlin butterknife

```kotlin
@file:Suppress("UNCHECKED_CAST", "DEPRECATION")

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.graphics.drawable.Drawable
import android.support.v4.app.Fragment
import android.support.v7.widget.RecyclerView
import android.view.View
import java.util.*
import kotlin.properties.ReadOnlyProperty
import kotlin.reflect.KProperty

class bind<in T, V : View>(val id: Int, val comment: String = "<No comment>", val rootView: View? = null) : ReadOnlyProperty<T, V> {

    var lazyValue: V? = null

    override fun getValue(thisRef: T, property: KProperty<*>): V {

        if (lazyValue != null) return lazyValue!!

        val injectView: V
        if (rootView != null) injectView = rootView.findViewById(id) as V
        else injectView = when (thisRef) {
            is Activity -> thisRef.findViewById(id) as V? ?: throw BindResourceError(id, "can't find view of $comment")

            is Fragment -> thisRef.view!!.findViewById(id) as V? ?: throw BindResourceError(id, "can't find view of $comment")

            is android.app.Fragment -> throw BindResourceError(id, "unknown rootView while bind $comment")

            is RecyclerView.ViewHolder -> thisRef.itemView.findViewById(id) as V? ?: throw BindResourceError(id, "can't find view of $comment")

            is ItemListAdapter.Holder<*> -> thisRef.rootView.findViewById(id) as V? ?: throw BindResourceError(id, "can't find view of $comment")

            is Dialog -> thisRef.findViewById(id) as V? ?: throw BindResourceError(id, "can't find view of $comment")

            is View -> thisRef.findViewById(id) as V? ?: throw BindResourceError(id, "can't find view of $comment")

            else -> throw BindResourceError(id, "unknown rootView while binding $comment")
        }
        lazyValue = injectView
        return injectView
    }
}

class bindViews<in T, V : View>(vararg val ids: Int, val comment: String = "<No comment>", val rootView: View? = null) : ReadOnlyProperty<T, List<V>> {

    var lazyValue: List<V>? = null

    override fun getValue(thisRef: T, property: KProperty<*>): List<V> {

        if (lazyValue != null) return lazyValue!!

        val result = ArrayList<V>()
        ids.forEach { id ->
            val injected: V = if (rootView != null)
                rootView.findViewById(id) as V
            else when (thisRef) {
                is Activity -> thisRef.findViewById(id) as V? ?: throw BindResourceError(id, "can't find view of $comment")

                is Fragment -> thisRef.view!!.findViewById(id) as V? ?: throw BindResourceError(id, "can't find view of $comment")

                is android.app.Fragment -> throw BindResourceError(id, "unknown rootView while bind $comment")

                is RecyclerView.ViewHolder -> thisRef.itemView.findViewById(id) as V? ?: throw BindResourceError(id, "can't find view of $comment")

                is Dialog -> thisRef.findViewById(id) as V? ?: throw BindResourceError(id, "can't find view of $comment")

                is View -> thisRef.findViewById(id) as V? ?: throw BindResourceError(id, "can't find view of $comment")

                else -> throw BindResourceError(id, "unknown rootView while binding $comment")
            }
            result.add(injected)
        }
        lazyValue = result
        return result
    }
}

class bindDrawable<in T, D : Drawable>(id: Int, comment: String = "<No comment>", rootContext: Context? = null) : bindResource<T, D>(id = id, comment = comment, rootContext = rootContext, getResource = { it.resources.getDrawable(id) as D })

class bindColor<in T>(id: Int, comment: String = "<No comment>", rootContext: Context? = null) : bindResource<T, Int>(id = id, comment = comment, rootContext = rootContext, getResource = { it.resources.getColor(id) })

class bindStr<in T>(id: Int, comment: String = "<No comment>", rootContext: Context? = null) : bindResource<T, String>(id = id, comment = comment, rootContext = rootContext, getResource = { it.getString(id) })

class bindInt<in T>(id: Int, comment: String = "<No comment>", rootContext: Context? = null) : bindResource<T, Int>(id = id, comment = comment, rootContext = rootContext, getResource = { it.resources.getInteger(id) })

class bindStrArray<in T>(id: Int, comment: String = "<No comment>", rootContext: Context? = null) : bindResource<T, Array<String>>(id = id, comment = comment, rootContext = rootContext, getResource = { it.resources.getStringArray(id) })

class bindDimension<in T>(id: Int, comment: String = "<No comment>", rootContext: Context? = null) : bindResource<T, Float>(id = id, comment = comment, rootContext = rootContext, getResource = { it.resources.getDimension(id) })

open class bindResource<in T, R>(val id: Int, val comment: String, val rootContext: Context? = null, val getResource: (Context) -> R?) : ReadOnlyProperty<T, R> {

    var lazyValue: R? = null

    override fun getValue(thisRef: T, property: KProperty<*>): R {

        if (lazyValue != null) return lazyValue!!

        return getResource(rootContext ?: thisRef!!.getSupportedContext() ?:
                throw BindResourceError(id, "unknown root context while bind $comment")) ?:
                throw BindResourceError(id, "can't find resource while bind $comment")
    }
}

class BindResourceError(id: Int, msg: String) : Exception("cant'bind resource $id for $msg")

inline fun <reified T : View> View.id(id: Int): T = findViewById(id) as T? ?: throw BindResourceError(id, "can't find view by id")

private fun Any.getSupportedContext(): Context? = when (this) {
    is Context -> this
    is View -> this.context
    is Fragment -> this.context
    is Dialog -> this.context
    is RecyclerView.ViewHolder -> this.itemView.context
    else -> App.getInstance()
}
```

you may use it like this

```
lass GroupCreateActivity : ToolbarActivityEx() {
    val mStepOneView: View by bind(R.id.activity_create_im_group_step_one, comment = "步骤一视图")
    val mStepTwoView: View by bind(R.id.activity_create_im_group_step_two, comment = "步骤二视图")
    val mIconView: StatusImageView by bind(R.id.activity_create_im_group_step_one_img, comment = "头像")
    val mNameEdit: EditText by bind(R.id.activity_create_im_group_step_two_name, comment = "群名")

...
```

**Comment by yperess** [comment]
Kotlin doesn't work with butterknife the example needs to be removed.

https://youtrack.jetbrains.com/issue/KT-18791

**Comment by j796160836** [comment]
I found change `annotationProcessor` to `kapt` in `build.gradle` works for me.

build.gradle

    dependencies {
        compile "com.jakewharton:butterknife:8.8.1"
        kapt "com.jakewharton:butterknife-compiler:8.8.1"
    }

and then you can do your familiar ButterKnife annotations 🙂

    class MainActivity : AppCompatActivity() {

        @BindView(R.id.myButton)
        lateinit var myButton: Button

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            setContentView(R.layout.activity_main)
            ButterKnife.bind(this)
            //...
        }
    }

**Comment by yperess** [comment]
@j796160836 This was fixed in 1.20-eap (which I'm guessing is what you're using). When it is merged into the stable channel this issue should be closed.
